File: sdks/android/src/main/kotlin/com/experimentation/sdk/ExperimentationSDK.kt
/**
 * Experimentation Platform — Android SDK
 *
 * Implements the Provider Abstraction pattern (ADR-007) for Kotlin/Android.
 *
 * Usage:
 *   val client = ExperimentClient(provider = RemoteProvider(baseUrl = "https://..."))
 *   val variant = client.getVariant("homepage_recs_v2", userId = "user-123")
 */
package com.experimentation.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

// ---------------------------------------------------------------------------
// Core Types
// ---------------------------------------------------------------------------

data class Assignment(
    val experimentId: String,
    val variantName: String,
    val payload: Map<String, Any> = emptyMap(),
    val fromCache: Boolean = false,
)

data class UserAttributes(
    val userId: String,
    val properties: Map<String, Any> = emptyMap(),
)

// ---------------------------------------------------------------------------
// Provider Interface
// ---------------------------------------------------------------------------

/**
 * All assignment backends implement this interface (ADR-007).
 */
interface AssignmentProvider {
    suspend fun initialize()
    suspend fun getAssignment(experimentId: String, attrs: UserAttributes): Assignment?
    suspend fun getAllAssignments(attrs: UserAttributes): Map<String, Assignment>
    suspend fun close()
}

// ---------------------------------------------------------------------------
// RemoteProvider
// ---------------------------------------------------------------------------

/**
 * Calls the Assignment Service via JSON HTTP.
 */
class RemoteProvider(
    private val baseUrl: String,
    private val timeoutMs: Long = 2000L,
) : AssignmentProvider {

    private var initialized = false

    override suspend fun initialize() {
        initialized = true
    }

    override suspend fun getAssignment(experimentId: String, attrs: UserAttributes): Assignment? {
        check(initialized) { "provider not initialized" }

        val url = "$baseUrl/experimentation.assignment.v1.AssignmentService/GetAssignment"
        val body = JSONObject().apply {
            put("userId", attrs.userId)
            put("experimentId", experimentId)
            put("attributes", JSONObject(attrs.properties.mapValues { (_, v) -> v.toString() }))
        }

        val responseJson = post(url, body) ?: return null
        if (!responseJson.optBoolean("isActive", false)) return null
        val variantId = responseJson.optString("variantId", "")
        if (variantId.isEmpty()) return null

        return Assignment(
            experimentId = responseJson.optString("experimentId", experimentId),
            variantName = variantId,
            payload = parsePayload(responseJson.optString("payloadJson", "")),
            fromCache = false,
        )
    }

    override suspend fun getAllAssignments(attrs: UserAttributes): Map<String, Assignment> {
        check(initialized) { "provider not initialized" }

        val url = "$baseUrl/experimentation.assignment.v1.AssignmentService/GetAssignments"
        val body = JSONObject().apply {
            put("userId", attrs.userId)
            put("attributes", JSONObject(attrs.properties.mapValues { (_, v) -> v.toString() }))
        }

        val responseJson = post(url, body) ?: return emptyMap()
        val assignmentsArray = responseJson.optJSONArray("assignments") ?: return emptyMap()

        val results = mutableMapOf<String, Assignment>()
        for (i in 0 until assignmentsArray.length()) {
            val a = assignmentsArray.getJSONObject(i)
            if (!a.optBoolean("isActive", false)) continue
            val variantId = a.optString("variantId", "")
            if (variantId.isEmpty()) continue
            val expId = a.getString("experimentId")
            results[expId] = Assignment(
                experimentId = expId,
                variantName = variantId,
                payload = parsePayload(a.optString("payloadJson", "")),
                fromCache = false,
            )
        }
        return results
    }

    override suspend fun close() {
        initialized = false
    }

    // -- Private --

    private suspend fun post(urlString: String, body: JSONObject): JSONObject? =
        withContext(Dispatchers.IO) {
            val conn = URL(urlString).openConnection() as HttpURLConnection
            try {
                conn.requestMethod = "POST"
                conn.connectTimeout = timeoutMs.toInt()
                conn.readTimeout = timeoutMs.toInt()
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true

                OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(body.toString()) }

                if (conn.responseCode != 200) return@withContext null

                val responseText = conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
                JSONObject(responseText)
            } finally {
                conn.disconnect()
            }
        }

    private fun parsePayload(jsonString: String): Map<String, Any> {
        if (jsonString.isEmpty()) return emptyMap()
        return try {
            val obj = JSONObject(jsonString)
            val result = mutableMapOf<String, Any>()
            for (key in obj.keys()) {
                result[key] = obj.get(key)
            }
            result
        } catch (_: Exception) {
            emptyMap()
        }
    }
}

// ---------------------------------------------------------------------------
// MockProvider
// ---------------------------------------------------------------------------

/**
 * Returns deterministic assignments for testing.
 */
class MockProvider(
    assignments: Map<String, String> = emptyMap(),
) : AssignmentProvider {

    private val mutex = Mutex()
    private val assignments = assignments.mapValues { (expId, variant) ->
        Assignment(experimentId = expId, variantName = variant)
    }.toMutableMap()

    override suspend fun initialize() {}

    override suspend fun getAssignment(experimentId: String, attrs: UserAttributes): Assignment? =
        mutex.withLock { assignments[experimentId] }

    override suspend fun getAllAssignments(attrs: UserAttributes): Map<String, Assignment> =
        mutex.withLock { assignments.toMap() }

    fun setAssignment(experimentId: String, variantName: String) {
        assignments[experimentId] = Assignment(experimentId = experimentId, variantName = variantName)
    }

    override suspend fun close() = mutex.withLock { assignments.clear() }
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

class ExperimentClient(
    private val provider: AssignmentProvider,
    private val fallback: AssignmentProvider? = null,
) {
    private var initialized = false

    suspend fun initialize() {
        provider.initialize()
        fallback?.initialize()
        initialized = true
    }

    suspend fun getVariant(
        experimentId: String,
        userId: String,
        properties: Map<String, Any> = emptyMap(),
    ): String? {
        if (!initialized) initialize()
        val attrs = UserAttributes(userId = userId, properties = properties)
        return try {
            provider.getAssignment(experimentId, attrs)?.variantName
        } catch (e: Exception) {
            fallback?.getAssignment(experimentId, attrs)?.variantName
        }
    }

    suspend fun close() {
        provider.close()
        fallback?.close()
        initialized = false
    }
}

File: sdks/ios/Sources/ExperimentationSDK.swift
// Experimentation Platform — iOS SDK
// Provider Abstraction pattern (ADR-007) for Swift.
//
// Usage:
//   let client = ExperimentClient(provider: RemoteProvider(baseURL: url))
//   let variant = try await client.getVariant("homepage_recs_v2", userId: "user-123")

import Foundation

// MARK: - Core Types

public struct Assignment: Sendable {
    public let experimentId: String
    public let variantName: String
    public let payload: [String: Any]
    public let fromCache: Bool

    public init(experimentId: String, variantName: String, payload: [String: Any] = [:], fromCache: Bool = false) {
        self.experimentId = experimentId
        self.variantName = variantName
        self.payload = payload
        self.fromCache = fromCache
    }
}

public struct UserAttributes: Sendable {
    public let userId: String
    public let properties: [String: String]

    public init(userId: String, properties: [String: String] = [:]) {
        self.userId = userId
        self.properties = properties
    }
}

// MARK: - Provider Protocol

/// All assignment backends implement this protocol (ADR-007).
public protocol AssignmentProvider: Sendable {
    func initialize() async throws
    func getAssignment(experimentId: String, attributes: UserAttributes) async throws -> Assignment?
    func getAllAssignments(attributes: UserAttributes) async throws -> [String: Assignment]
    func close() async
}

// MARK: - RemoteProvider

/// Calls the Assignment Service via JSON HTTP.
public final class RemoteProvider: AssignmentProvider {
    private let baseURL: URL
    private let timeoutSeconds: TimeInterval
    private var session: URLSession?

    public init(baseURL: URL, timeoutSeconds: TimeInterval = 2.0) {
        self.baseURL = baseURL
        self.timeoutSeconds = timeoutSeconds
    }

    public func initialize() async throws {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeoutSeconds
        session = URLSession(configuration: config)
    }

    public func getAssignment(experimentId: String, attributes: UserAttributes) async throws -> Assignment? {
        let url = baseURL.appendingPathComponent(
            "experimentation.assignment.v1.AssignmentService/GetAssignment"
        )
        let body: [String: Any] = [
            "userId": attributes.userId,
            "experimentId": experimentId,
            "attributes": attributes.properties,
        ]
        let data = try await post(url: url, body: body)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        guard json["isActive"] as? Bool == true,
              let variantId = json["variantId"] as? String,
              !variantId.isEmpty else {
            return nil
        }
        let experimentIdResp = json["experimentId"] as? String ?? experimentId
        let payloadJson = json["payloadJson"] as? String ?? ""
        var payload: [String: Any] = [:]
        if !payloadJson.isEmpty,
           let d = payloadJson.data(using: .utf8),
           let obj = try? JSONSerialization.jsonObject(with: d) as? [String: Any] {
            payload = obj
        }
        return Assignment(
            experimentId: experimentIdResp,
            variantName: variantId,
            payload: payload,
            fromCache: false
        )
    }

    public func getAllAssignments(attributes: UserAttributes) async throws -> [String: Assignment] {
        let url = baseURL.appendingPathComponent(
            "experimentation.assignment.v1.AssignmentService/GetAssignments"
        )
        let body: [String: Any] = [
            "userId": attributes.userId,
            "attributes": attributes.properties,
        ]
        let data = try await post(url: url, body: body)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let assignments = json["assignments"] as? [[String: Any]] else {
            return [:]
        }
        var results: [String: Assignment] = [:]
        for a in assignments {
            guard a["isActive"] as? Bool == true,
                  let variantId = a["variantId"] as? String,
                  !variantId.isEmpty,
                  let expId = a["experimentId"] as? String else {
                continue
            }
            let payloadJson = a["payloadJson"] as? String ?? ""
            var payload: [String: Any] = [:]
            if !payloadJson.isEmpty,
               let d = payloadJson.data(using: .utf8),
               let obj = try? JSONSerialization.jsonObject(with: d) as? [String: Any] {
                payload = obj
            }
            results[expId] = Assignment(
                experimentId: expId,
                variantName: variantId,
                payload: payload,
                fromCache: false
            )
        }
        return results
    }

    public func close() async {
        session?.invalidateAndCancel()
        session = nil
    }

    // MARK: - Private

    private func post(url: URL, body: [String: Any]) async throws -> Data {
        guard let session else {
            throw NSError(domain: "ExperimentationSDK", code: -1, userInfo: [
                NSLocalizedDescriptionKey: "provider not initialized"
            ])
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        guard let httpResp = response as? HTTPURLResponse, httpResp.statusCode == 200 else {
            return Data()
        }
        return data
    }
}

// MARK: - MockProvider

/// Returns deterministic assignments for testing.
public final class MockProvider: AssignmentProvider {
    private var assignments: [String: Assignment] = [:]

    public init(assignments: [String: String] = [:]) {
        for (expId, variant) in assignments {
            self.assignments[expId] = Assignment(experimentId: expId, variantName: variant)
        }
    }

    public func initialize() async throws {}

    public func getAssignment(experimentId: String, attributes: UserAttributes) async throws -> Assignment? {
        return assignments[experimentId]
    }

    public func getAllAssignments(attributes: UserAttributes) async throws -> [String: Assignment] {
        return assignments
    }

    public func setAssignment(experimentId: String, variantName: String) {
        assignments[experimentId] = Assignment(experimentId: experimentId, variantName: variantName)
    }

    public func close() async {
        assignments.removeAll()
    }
}

// MARK: - Client

public final class ExperimentClient {
    private let provider: AssignmentProvider
    private let fallback: AssignmentProvider?

    public init(provider: AssignmentProvider, fallback: AssignmentProvider? = nil) {
        self.provider = provider
        self.fallback = fallback
    }

    public func initialize() async throws {
        try await provider.initialize()
        try await fallback?.initialize()
    }

    public func getVariant(_ experimentId: String, userId: String, properties: [String: String] = [:]) async throws -> String? {
        let attrs = UserAttributes(userId: userId, properties: properties)
        do {
            return try await provider.getAssignment(experimentId: experimentId, attributes: attrs)?.variantName
        } catch {
            return try await fallback?.getAssignment(experimentId: experimentId, attributes: attrs)?.variantName
        }
    }

    public func close() async {
        await provider.close()
        await fallback?.close()
    }
}
